### ISSUE TITLE
batch inference

### ISSUE THREAD

**Issue by khanhha** [open-mmlab/mmpose#25]
Hello,
I am trying to find a way to do batch inference with mmpose, but it seems that the project has not supported it yet. There is only a function called "**_inference_single_pose_model**" inside inference.py.

Is there any quick way I can adapt mmpose to batch inference? I think it's an essential feature because most of the time, we just work on bach inference. real-time infernece with a single input image is not possible with the current architectures, so single-image inference is kind of not useful.

Thank  you and I look forward to hearing from you soon.
Best.

**Comment by jin-s13** [comment]
Thanks for the suggestion. We will add this feature soon.

As a workaround, you can customize a dataset and test the dataset.  
see the `Test a dataset` section, in `docs/getting_started.md`

**Comment by khanhha** [comment]
Thank you very much!

**Comment by khanhha** [comment]
here is my solution so far. Hope that it will be helpful for others

    def estimate_batch(self, imgs: List[np.ndarray]):
        """
        :param imgs: Nx RGB images
        :return: Nx17x3. [x,y,c] c means confidence score
        """
        model_input_w, model_input_h = self.input_img_size
        resize_scales = [(img.shape[1] / model_input_w, img.shape[0] / model_input_h) for img in imgs]

        trans = self.input_transforms
        self.model.eval()
        with torch.no_grad():
            inputs = torch.cat([trans(Image.fromarray(img)).unsqueeze(0) for img in imgs], dim=0).to(self.device)
            output = self.model.backbone(inputs)
            output = self.model.keypoint_head(output)
            preds, maxvals = _keypoints_from_heatmaps(output.clone().cpu().numpy(),
                                                      post_process=self.model.test_cfg['post_process'],
                                                      unbiased=True,
                                                      kernel=self.model.test_cfg['modulate_kernel'])
            heatmap_width, heatmap_height = self.heatmap_size
            scales = [model_input_w / heatmap_width, model_input_h / heatmap_height]
            preds[:, :, 0] *= scales[0]
            preds[:, :, 1] *= scales[1]
            poses = np.concatenate([preds, maxvals], axis=2)
            for i in range(len(poses)):
                poses[i, :, 0] *= resize_scales[i][0]
                poses[i, :, 1] *= resize_scales[i][1]

        return poses

**Comment by jin-s13** [comment]
@khanhha Could you please offer some help? #781

**Comment by denfromufa** [comment]
What is “self” object in the function above? @khanhha @jin-s13

**Comment by jin-s13** [comment]
> What is “self” object in the function above? @khanhha @jin-s13

https://www.geeksforgeeks.org/self-in-python-class/#:~:text=self%20represents%20the%20instance%20of,to%20refer%20to%20instance%20attributes.

**Comment by tlemangen** [comment]
> here is my solution so far. Hope that it will be helpful for others这是我目前的解决方案。希望对其他人有所帮助
> 
> ```
> def estimate_batch(self, imgs: List[np.ndarray]):
>     """
>     :param imgs: Nx RGB images
>     :return: Nx17x3. [x,y,c] c means confidence score
>     """
>     model_input_w, model_input_h = self.input_img_size
>     resize_scales = [(img.shape[1] / model_input_w, img.shape[0] / model_input_h) for img in imgs]
> 
>     trans = self.input_transforms
>     self.model.eval()
>     with torch.no_grad():
>         inputs = torch.cat([trans(Image.fromarray(img)).unsqueeze(0) for img in imgs], dim=0).to(self.device)
>         output = self.model.backbone(inputs)
>         output = self.model.keypoint_head(output)
>         preds, maxvals = _keypoints_from_heatmaps(output.clone().cpu().numpy(),
>                                                   post_process=self.model.test_cfg['post_process'],
>                                                   unbiased=True,
>                                                   kernel=self.model.test_cfg['modulate_kernel'])
>         heatmap_width, heatmap_height = self.heatmap_size
>         scales = [model_input_w / heatmap_width, model_input_h / heatmap_height]
>         preds[:, :, 0] *= scales[0]
>         preds[:, :, 1] *= scales[1]
>         poses = np.concatenate([preds, maxvals], axis=2)
>         for i in range(len(poses)):
>             poses[i, :, 0] *= resize_scales[i][0]
>             poses[i, :, 1] *= resize_scales[i][1]
> 
>     return poses
> ```

I'm new to mmpose, and I've noticed that my `self.model` doesn't have a `keypoint_head` method. I also don't know where to find the `_keypoints_from_heatmaps` method. Could you help me figure this out? I'd really appreciate it!
